### PR TITLE
Bugfix FXIOS-9406 Fix intermittent blank homepage 

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9406)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20831)

## :bulb: Description

I believe our use of `actor` for TabSessionStore is incorrect for a few reasons, and also problematic because it forces us to use an async `Task` in our `selectTab` function which should effectively be synchronous.

The change in this PR is to remove the use of `actor` and instead leverage a simple mutex to ensure atomicity of the on-disk files. Offhand I don't believe this lock is actually necessary 🤔 , however for now it will ensure that any concerns around concurrent reads/writes to the files are alleviated, and the lock should be extremely fast and contention should generally be minimal (or non-existent).

In practice, removing this use of `Task` appears to fix the blank homepage issue for me reliably. I also can't observe any issues with the back navigation button as we observed for the 8/7 incident. (This general area of the codebase still needs further evaluation and updates, we may for example be able to remove the workaround we originally added for the back button now that this removal of the actor is in place.)

## Unit Tests

To keep the PR diff readable, I haven't yet updated unit tests etc. for this refactor. Still forthcoming.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by dataports** [comment]
Can't repro any of the issues with homepage and did a bunch of checks around opening and closing the apps to make sure the tabs were persisted. Looks good to me.

**Review by dataports — APPROVED** [review]

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:438]: Nice! Can we add another assert to `BrowserViewController.swift` line 3470 to check that the `selected` Tab has a non-nil `webView` as well? That caused all the other problems we saw.

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:367]: Omg, love this fix!! 🔥 Saving/loading the `webView`'s `interactionState` asynchronicity was the source of several bugs.

Can you also clear the extra `didSelectedTabChange` delegate call that is currently in `selectTabWithSession` (documented as `FXIOS-9785 Note 1`)? We shouldn't need the duplicate the call anymore since we don't have to await this Task. But we should test that too just to make sure.

You could probably do a quick search for `FXIOS-9785` and clear the other comments I made as this should address everything now!

File: BrowserKit/Sources/TabDataStore/TabSessionStore.swift
Comment (ih-codes) [inline BrowserKit/Sources/TabDataStore/TabSessionStore.swift:36]: Because of this, `saveSessionData` method no longer needs a Task await on line 312 of `TabManagerImplementation.swift`.

**Review by ih-codes — COMMENTED** [review]
Loving this fix! 🎉  It should prevent so many other future bugs too, that async save/load of the webView `interactionState` was causing us grief on at least 3 other tickets I've looked at, not to mention test "memory leak" errors.

FYI some unit tests need to be updated now because the awaits are removed. I could do that for you if you want if I have permissions to push to your branch.

I'll run some more tests after any changes you make from my review before I approve.

**Comment by mattreaganmozilla** [comment]
> FYI some unit tests need to be updated now because the awaits are removed

@ih-codes Yeah absolutely, I had planned to push those up later in order to keep the PR diff minimal but I should be able to fix those fairly quickly.

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (mattreaganmozilla) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:367]: > Can you also clear the extra didSelectedTabChange delegate call that is currently in selectTabWithSession (documented as FXIOS-9785 Note 1)?

Yeah I wondered if we should remove that immediately. I will remove it and do another round of testing and unless there's a problem I'll push that up shortly.

File: BrowserKit/Sources/TabDataStore/TabSessionStore.swift
Comment (mattreaganmozilla) [reply BrowserKit/Sources/TabDataStore/TabSessionStore.swift:36]: Thank you, fixed

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:365]: Yeah if you push that up I'll definitely re-test the incident situation and try to test the white screen issue as well.

@dataports Maybe you can test on a device as well (for the white screen bug) once @mattreaganmozilla pushes up this change (reverting back to your initial fix for it).

Comment (mattreaganmozilla) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:365]: I've pushed that up, I can't see any issues so far with it removed 👍

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:365]: Awesome! 👍  I will test some more and approve if I don't find anything else! (I can only test sim though)

**Review by ih-codes — APPROVED** [review]
Tested on iPhone/iPad sims, no issues with backButton enabled or toolBar scrolling to hide (incident issues). Also couldn't reproduce on the sim (even with my breakpoint) the blank homepage bug. 👍

**Comment by mattreaganmozilla** [comment]
Going to add Mergify for 130 & 129, per @DonalMe:

> Doesn't stop getting the fix in main when it's ready and merged to release/v130. A backport for release/v129can be created but not merged, relman will handle that when we prep the weekly release.

**Comment by mattreaganmozilla** [comment]
@Mergifyio backport release/v130 release/v129

**Pull request merged**
